src/vast_tracker.js

```diff
@@ -368,84 +376,125 @@ export class VASTTracker extends EventEmitter {
-    this.track('adExpand');
+  adExpand(macros = {}) {
+    this.track('adExpand', macros);
   }
 
   /**
    * Must be called when the user activated a control to reduce the creative to its original dimensions.
    * Calls the adCollapse tracking URLs.
    *
+   * @param {Object} [macros={}] - An optional Object containing macros and their values to be used and replaced in the tracking calls.
    * @emits VASTTracker#adCollapse
    */
-  adCollapse() {
-    this.track('adCollapse');
+  adCollapse(macros = {}) {
+    this.track('adCollapse', macros);
   }
 
   /**
    * Must be called if the user clicked or otherwise activated a control used to minimize the ad.
    * Calls the minimize tracking URLs.
+   *
+   * @param {Object} [macros={}] - An optional Object containing macros and their values to be used and replaced in the tracking calls.
    * @emits VASTTracker#minimize
    */
-  minimize() {
-    this.track('minimize');
+  minimize(macros = {}) {
+    this.track('minimize', macros);
+  }
+
+  /**
+   * Must be called if the player did not or was not able to execute the provided
+   * verification code.The [REASON] macro must be filled with reason code
+   * Calls the verificationNotExecuted trackings URLs.
+   *
+   * @param {Object} [macros={}] - An optional Object containing macros and their values to be used and replaced in the tracking calls.
+   * @emits VASTTracker#verificationNotExecuted
+   */
+  verificationNotExecuted(macros = {}) {
+    if (
+      !(this.ad && this.ad.adVerifications && this.ad.adVerifications.length)
+    ) {
+      return;
+    }
+
+    this.ad.adVerifications.forEach(verification => {
+      if (
+        verification.trackingEvents &&
+        verification.trackingEvents.verificationNotExecuted
+      ) {
+        this.trackURLs(
+          verification.trackingEvents.verificationNotExecuted,
+          macros
+        );
+        this.emit('verificationNotExecuted', null);
+      }
+    });
   }
 
   /**
    * The time that the initial ad is displayed. This time is based on
    * the time between the impression and either the completed length of display based
    * on the agreement between transactional parties or a close, minimize, or accept
    * invitation event.
-   * The time has to be passed using [ADPLAYHEAD] and [MEDIAPLAYHEAD] for VAST 4.1
+   * The time will be passed using [ADPLAYHEAD] and [MEDIAPLAYHEAD] macros for VAST 4.1
    * Calls the overlayViewDuration tracking URLs.
+   *
+   * @param {String} duration - The time that the initial ad is displayed.
+   * @param {Object} [macros={}] - An optional Object containing macros and their values to be used and replaced in the tracking calls.
    * @emits VASTTracker#overlayViewDuration
    */
-  // TODO : Pass the duration as a parameter and replace the corresponding macro with it
-  overlayViewDuration() {
-    this.track('overlayViewDuration');
+  overlayViewDuration(duration, macros = {}) {
+    macros['CONTENTPLAYHEAD'] = duration;
```

Comment:
btw shouldn't all these macros be encoded with `encodeURIComponentRFC3986`?

Reply:
They are now all encoded